### PR TITLE
event: Fix a crash by adding PROT_READ along with PROT_EXEC

### PR DESCRIPTION
**Pull request by honggyukim**
Currently, uftrace gets segfault when patching SDT events into the
target binary.
```
  $ gcc -pg -o t-sdt s-sdt.c

  $ uftrace -E uftrace:* --match glob t-sdt
  WARN: child terminated by signal: 11: Segmentation fault
  WARN: cannot open record data: /tmp/uftrace-live-CCoVbF: No data available
```
The reason of the segfault is because the access permission of the
SDT memory region becomes PROT_EXEC only without having PROT_READ.

It might be okay in some environments but some are not so this patch
explicitly adds PROT_READ along with PROT_EXEC.

Before:
```
  147 event_sdt           : SG SG SG SG SG SG SG SG SG SG
  150 recv_event          : NZ NZ NZ NZ NZ NZ NZ NZ NZ NZ
```
After:
```
  147 event_sdt           : OK OK OK OK OK OK OK OK OK OK
  150 recv_event          : OK OK OK OK OK OK OK OK OK OK
```
Fixed: #1286

Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>

### CODE REVIEW

File: utils/fstack.c [hunk 1]
```diff
@@ -1045,7 +1045,7 @@ bool fstack_check_opts(struct uftrace_task_reader *task, struct opts *opts)
 			return false;
 	}
 
-	if (opts->no_event && rec->type == UFTRACE_EVENT)
+	if (opts->no_event && !opts->event && rec->type == UFTRACE_EVENT)
```
Comment (namhyung) [inline utils/fstack.c:1048]: This looks like a separate fix.  Can you please split into another commit?

Comment (honggyukim) [reply utils/fstack.c:1048]: OK. I just split it into a new commit.

File: arch/x86_64/mcount-event.c [hunk 2]
```diff
@@ -50,10 +50,11 @@ int mcount_arch_enable_event(struct mcount_event_info *mei)
 		return -1;
 	}
 
-	/* replace NOP to an invalid OP so that it can catch SIGILL */
+	/* replace NOP to an invalid OP so that it can catch SIGILL,
+	   then it will fall into sdt_handler() above. */
 	memset((void *)mei->addr, INVALID_OPCODE, 1);
 
-	if (mprotect(PAGE_ADDR(mei->addr), PAGE_SIZE, PROT_EXEC))
+	if (mprotect(PAGE_ADDR(mei->addr), PAGE_SIZE, PROT_READ | PROT_EXEC))
```
Comment (namhyung) [inline arch/x86_64/mcount-event.c:57]: Could you please take a look at the other callsites of mproctect() and fix them all?

Comment (honggyukim) [reply arch/x86_64/mcount-event.c:57]: I also fixed it in i386 as well.

**Comment by honggyukim** [comment]
I found that we can enable `-mfentry -mnop-mcount` build in i386 as well so made it work together.

**Comment by namhyung** [comment]
Oh, good. But I don't want to put it here together.  It's a separate change and would belong to another PR.

**Comment by honggyukim** [comment]
I put this together because the i386 patch requires `uftrace: Add PROT_READ permission along with PROT_EXEC` but any of them needs to be merged first. Otherwise, I have to put the all together. Same for the test fix PR as well. As the PR grows, some of the patches are needed together so any of the PR needs to be merged first.

**Comment by honggyukim** [comment]
I removed `i386` changes in #1296 separately.

**Review by namhyung — APPROVED** [review]
LGTM

**Comment by honggyukim** [comment]
I will rebase other PRs.
